ocaml/testsuite/tests/comprehensions/comprehensions.ml

```diff
@@ -0,0 +1,554 @@
+ (2, 4, 0); (4, 4, 0); (0, 0, 2); (2, 0, 2); (4, 0, 2); (0, 2, 2); (2, 2, 2);
+ (4, 2, 2); (0, 4, 2); (2, 4, 2); (4, 4, 2); (0, 0, 4); (2, 0, 4); (4, 0, 4);
+ (0, 2, 4); (2, 2, 4); (4, 2, 4); (0, 4, 4); (2, 4, 4); (4, 4, 4)]
+|}];;
+
+[| (i,j,k) for i = 0 to 5  and j = 0 to 5 and k = 0 to 5 when ((k mod 2 = 0) && (i mod 2 = 0) && (j mod 2 = 0)  )|];;
+[%%expect{|
+- : (int * int * int) array =
+[|(0, 0, 0); (2, 0, 0); (4, 0, 0); (0, 2, 0); (2, 2, 0); (4, 2, 0);
+  (0, 4, 0); (2, 4, 0); (4, 4, 0); (0, 0, 2); (2, 0, 2); (4, 0, 2);
+  (0, 2, 2); (2, 2, 2); (4, 2, 2); (0, 4, 2); (2, 4, 2); (4, 4, 2);
+  (0, 0, 4); (2, 0, 4); (4, 0, 4); (0, 2, 4); (2, 2, 4); (4, 2, 4);
+  (0, 4, 4); (2, 4, 4); (4, 4, 4)|]
+|}];;
+
+[| (i,j,k) for i = 0 to 5 when ((k mod 2 = 0) && (i mod 2 = 0) && (j mod 2 = 0)) for j = 0 to 5 for k = 0 to 5 |];;
+[%%expect{|
+- : (int * int * int) array =
+[|(0, 0, 0); (2, 0, 0); (4, 0, 0); (0, 2, 0); (2, 2, 0); (4, 2, 0);
+  (0, 4, 0); (2, 4, 0); (4, 4, 0); (0, 0, 2); (2, 0, 2); (4, 0, 2);
+  (0, 2, 2); (2, 2, 2); (4, 2, 2); (0, 4, 2); (2, 4, 2); (4, 4, 2);
+  (0, 0, 4); (2, 0, 4); (4, 0, 4); (0, 2, 4); (2, 2, 4); (4, 2, 4);
+  (0, 4, 4); (2, 4, 4); (4, 4, 4)|]
+|}];;
+
+let f f =
+  [ (f i j k) for i = 0 to 3 when (i mod 2 = 0)  for j = 0 to 3 when (j mod 2 = 0)  for k = 0 to 3 when (k mod 2 = 0)];;
+[%%expect{|
+val f : (int -> int -> int -> 'a) -> 'a list = <fun>
+|}];;
+
+f (fun i j k -> (i,j,k) )
+[%%expect{|
+- : (int * int * int) list =
+[(0, 0, 0); (2, 0, 0); (0, 2, 0); (2, 2, 0); (0, 0, 2); (2, 0, 2); (0, 2, 2);
+ (2, 2, 2)]
+|}];;
+
+f (fun i j k -> i )
+[%%expect{|
+- : int list = [0; 2; 0; 2; 0; 2; 0; 2]
+|}];;
+
+f (fun i j k -> float_of_int i )
+[%%expect{|
+- : float list = [0.; 2.; 0.; 2.; 0.; 2.; 0.; 2.]
+|}];;
+
+f (fun i j k -> [|string_of_int i|] )
+[%%expect{|
+- : string array list =
+[[|"0"|]; [|"2"|]; [|"0"|]; [|"2"|]; [|"0"|]; [|"2"|]; [|"0"|]; [|"2"|]]
+|}];;
+
+
+(*Make sure stuff is called in correct order/ correct number of times.*)
+
+let var = ref [];;
+let f x = var := x::!var; x;;
+[%%expect{|
+val var : '_weak1 list ref = {contents = []}
+val f : '_weak2 -> '_weak2 = <fun>
+|}];;
+
+let z = [|1;2;3|];;
+[| i  for  i in (f z) |];;
+List.rev !var;;
+[%%expect{|
+val z : int array = [|1; 2; 3|]
+- : int array = [|1; 2; 3|]
+- : int array list = [[|1; 2; 3|]]
+|}];;
+
+var := [];;
+let z = [|1;2;3|];;
+[| i  for  i in (f z) for i = 0 to 1 |];;
+List.rev !var;;
+[%%expect{|
+- : unit = ()
+val z : int array = [|1; 2; 3|]
+- : int array = [|1; 2; 3; 1; 2; 3|]
+- : int array list = [[|1; 2; 3|]; [|1; 2; 3|]]
+|}];;
+
+var := [];;
+let z = [|1;2;3|];;
+[| i  for  i in (f z) for i = 0 to 6 when (i mod 2 = 0) |];;
+List.rev !var;;
+[%%expect{|
+- : unit = ()
+val z : int array = [|1; 2; 3|]
+- : int array = [|1; 2; 3; 1; 2; 3; 1; 2; 3; 1; 2; 3|]
+- : int array list = [[|1; 2; 3|]; [|1; 2; 3|]; [|1; 2; 3|]; [|1; 2; 3|]]
+|}];;
+
+var := [];;
+let z = [|1;2;3|];;
+[| i  for  i in (f z) and i = 0 to 1 |];;
```

Comment:
As it turns out this here gives an unused warning:

`[| i  for  i in (f z) and i in (f z) |];;`

The one in the tests does not, since it uses the same typing as the for loop, which should give a `Unused_for_index` warning. I guess the `Unused_for_index` warning is on a different level and needs to be enabled, doesn't it?